### PR TITLE
[WIP] add validation for address orders to prevent exceptions (type hi…

### PR DESCRIPTION
**Pull request by healerz**
…nts) in case of missing properties provided by the client and to stick to the command bus pattern

### CODE REVIEW

File: src/Request/AddressRequest.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /** @var string|null */
+    private $provinceCode;
+
+    /** @var string|null */
+    private $company;
+
+    /** @var string|null */
+    private $phoneNumber;
+
+    public function __construct($address) {
```
Comment (mamazu) [inline src/Request/AddressRequest.php]: Missing a typehint

File: src/Request/AddressOrderRequest.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /** @var AddressRequest */
+    private $address;
+
+    /** @var AddressRequest */
+    private $billingAddress;
+
+    public function __construct(Request $request)
+    {
+        $this->orderToken = $request->attributes->get('token');
+        $this->address = new AddressRequest($request->request->get('shippingAddress'));
```
Comment (mamazu) [inline src/Request/AddressOrderRequest.php]: We also need to validate that `$request->request->get('shippingAddress')` is not null.

**Comment by healerz** [comment]
just in case the context of this PR is not clear, here is the related issue i created: https://github.com/Sylius/ShopApiPlugin/issues/353

File: src/Request/AddressRequest.php (hunk 1 above)
Comment (healerz) [reply src/Request/AddressRequest.php]: Actually, I didn't type hint the param on purpose. As I see it, the request objects are just loose data holder objects, which need to be validated by the validator anyway and they shouldn't end in an exception in any case.
But you are right, in this case I could type hint it with "?array". On the other hand, if I finish the whole PR, i also need to adjust the code for the Address actions and then i need to modify the Address constructor to something that fits for both cases anyway.

Comment (mamazu) [reply src/Request/AddressRequest.php]: Then if `$address` is null you can not use the array access to get properties.

File: src/Request/AddressOrderRequest.php (hunk 2 above)
Comment (mamazu) [reply src/Request/AddressOrderRequest.php]: Or you could add a default to it that it returns an empty array. Like so: `$request->request->get('shippingAddress', [])`

File: src/Request/AddressRequest.php (hunk 1 above)
Comment (healerz) [reply src/Request/AddressRequest.php]: with the `??` operator i can ;)

> php > $foo = null;
php > var_dump($foo['bar'] ?? null);
NULL
php > $foo['bar'] = 'bar';
php > var_dump($foo['bar'] ?? null);
string(3) "bar"

File: src/Request/AddressOrderRequest.php (hunk 2 above)
Comment (healerz) [reply src/Request/AddressOrderRequest.php]: yes i could :) The behaviour is in either way the same

File: src/Request/AddressRequest.php (hunk 1 above)
Comment (mamazu) [reply src/Request/AddressRequest.php]: Ah okay. Didn't expect that to work. :D

Comment (healerz) [reply src/Request/AddressRequest.php]: The magic of the `??` operator :D
It's just the shorthand of `isset($foo['bar']) ? $foo['bar] : null`

**Comment by mamazu** [comment]
The idea looks good. This also allows for better error handling in the API client.

**Comment by mamazu** [comment]
If you fix the tests. This feature would be ready for release I'd say.

**Comment by mamazu** [comment]
This should be closed by #453. Isn't it? @healerz

**Comment by lchrusciel** [comment]
Hey Michel, 

please reopen if still relevant :)

**Pull request closed without merging**